// drizzle-orm/d1/driver.js
import * as V1 from "../_relations.js";
import { entityKind } from "../entity.js";
import { DefaultLogger } from "../logger.js";
import { BaseSQLiteDatabase } from "../sqlite-core/db.js";
import { SQLiteAsyncDialect } from "../sqlite-core/dialect.js";
import { SQLiteD1Session } from "./session.js";
class DrizzleD1Database extends BaseSQLiteDatabase {
  static [entityKind] = "D1Database";
  async batch(batch) {
    return this.session.batch(batch);
  }
}
function drizzle(client, config = {}) {
  const dialect = new SQLiteAsyncDialect({ casing: config.casing });
  let logger;
  if (config.logger === true) {
    logger = new DefaultLogger();
  } else if (config.logger !== false) {
    logger = config.logger;
  }
  let schema;
  if (config.schema) {
    const tablesConfig = V1.extractTablesRelationalConfig(
      config.schema,
      V1.createTableRelationsHelpers
    );
    schema = {
      fullSchema: config.schema,
      schema: tablesConfig.tables,
      tableNamesMap: tablesConfig.tableNamesMap
    };
  }
  const relations = config.relations;
  const session = new SQLiteD1Session(client, dialect, relations, schema, { logger });
  const db = new DrizzleD1Database(
    "async",
    dialect,
    session,
    relations,
    schema,
    void 0,
    true
  );
  db.$client = client;
  return db;
}
export {
  DrizzleD1Database,
  drizzle
};
