// drizzle-orm/gel-core/query-builders/query.js
import { entityKind } from "../../entity.js";
import { QueryPromise } from "../../query-promise.js";
import {
  mapRelationalRow
} from "../../relations.js";
import { tracer } from "../../tracing.js";
class RelationalQueryBuilder {
  constructor(tables, schema, tableNamesMap, table, tableConfig, dialect, session) {
    this.tables = tables;
    this.schema = schema;
    this.tableNamesMap = tableNamesMap;
    this.table = table;
    this.tableConfig = tableConfig;
    this.dialect = dialect;
    this.session = session;
  }
  static [entityKind] = "GelRelationalQueryBuilderV2";
  findMany(config) {
    return new PgRelationalQuery(
      this.tables,
      this.schema,
      this.tableNamesMap,
      this.table,
      this.tableConfig,
      this.dialect,
      this.session,
      config ?? true,
      "many"
    );
  }
  findFirst(config) {
    return new PgRelationalQuery(
      this.tables,
      this.schema,
      this.tableNamesMap,
      this.table,
      this.tableConfig,
      this.dialect,
      this.session,
      config ?? true,
      "first"
    );
  }
}
class PgRelationalQuery extends QueryPromise {
  constructor(tables, schema, tableNamesMap, table, tableConfig, dialect, session, config, mode) {
    super();
    this.tables = tables;
    this.schema = schema;
    this.tableNamesMap = tableNamesMap;
    this.table = table;
    this.tableConfig = tableConfig;
    this.dialect = dialect;
    this.session = session;
    this.config = config;
    this.mode = mode;
  }
  static [entityKind] = "GelRelationalQueryV2";
  /** @internal */
  _prepare(name) {
    return tracer.startActiveSpan("drizzle.prepareQuery", () => {
      const { query, builtQuery } = this._toSQL();
      return this.session.prepareRelationalQuery(
        builtQuery,
        void 0,
        name,
        (rawRows, mapColumnValue) => {
          const rows = rawRows.map((row) => mapRelationalRow(row, query.selection, mapColumnValue));
          if (this.mode === "first") {
            return rows[0];
          }
          return rows;
        }
      );
    });
  }
  prepare(name) {
    return this._prepare(name);
  }
  _getQuery() {
    return this.dialect.buildRelationalQuery({
      schema: this.schema,
      tableNamesMap: this.tableNamesMap,
      table: this.table,
      tableConfig: this.tableConfig,
      queryConfig: this.config,
      tables: this.tables,
      mode: this.mode
    });
  }
  /** @internal */
  getSQL() {
    return this._getQuery().sql;
  }
  _toSQL() {
    const query = this._getQuery();
    const builtQuery = this.dialect.sqlToQuery(query.sql);
    return { query, builtQuery };
  }
  toSQL() {
    return this._toSQL().builtQuery;
  }
  authToken;
  /** @internal */
  setToken(token) {
    this.authToken = token;
    return this;
  }
  execute() {
    return tracer.startActiveSpan("drizzle.operation", () => {
      return this._prepare().execute();
    });
  }
}
export {
  PgRelationalQuery,
  RelationalQueryBuilder
};
