// drizzle-orm/sqlite-core/query-builders/query.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var query_exports = {};
__export(query_exports, {
  RelationalQueryBuilder: () => RelationalQueryBuilder,
  SQLiteRelationalQuery: () => SQLiteRelationalQuery,
  SQLiteSyncRelationalQuery: () => SQLiteSyncRelationalQuery
});
module.exports = __toCommonJS(query_exports);
var import_entity = require("../../entity.cjs");
var import_query_promise = require("../../query-promise.cjs");
var import_relations = require("../../relations.cjs");
var import_sql = require("../../sql/sql.cjs");
class RelationalQueryBuilder {
  constructor(mode, tables, schema, tableNamesMap, table, tableConfig, dialect, session, rowMode, forbidJsonb) {
    this.mode = mode;
    this.tables = tables;
    this.schema = schema;
    this.tableNamesMap = tableNamesMap;
    this.table = table;
    this.tableConfig = tableConfig;
    this.dialect = dialect;
    this.session = session;
    this.rowMode = rowMode;
    this.forbidJsonb = forbidJsonb;
  }
  static [import_entity.entityKind] = "SQLiteAsyncRelationalQueryBuilderV2";
  findMany(config) {
    return this.mode === "sync" ? new SQLiteSyncRelationalQuery(
      this.tables,
      this.schema,
      this.tableNamesMap,
      this.table,
      this.tableConfig,
      this.dialect,
      this.session,
      config ?? true,
      "many",
      this.rowMode,
      this.forbidJsonb
    ) : new SQLiteRelationalQuery(
      this.tables,
      this.schema,
      this.tableNamesMap,
      this.table,
      this.tableConfig,
      this.dialect,
      this.session,
      config ?? true,
      "many",
      this.rowMode,
      this.forbidJsonb
    );
  }
  findFirst(config) {
    return this.mode === "sync" ? new SQLiteSyncRelationalQuery(
      this.tables,
      this.schema,
      this.tableNamesMap,
      this.table,
      this.tableConfig,
      this.dialect,
      this.session,
      config ?? true,
      "first",
      this.rowMode,
      this.forbidJsonb
    ) : new SQLiteRelationalQuery(
      this.tables,
      this.schema,
      this.tableNamesMap,
      this.table,
      this.tableConfig,
      this.dialect,
      this.session,
      config ?? true,
      "first",
      this.rowMode,
      this.forbidJsonb
    );
  }
}
class SQLiteRelationalQuery extends import_query_promise.QueryPromise {
  constructor(tables, schema, tableNamesMap, table, tableConfig, dialect, session, config, mode, rowMode, forbidJsonb) {
    super();
    this.tables = tables;
    this.schema = schema;
    this.tableNamesMap = tableNamesMap;
    this.tableConfig = tableConfig;
    this.dialect = dialect;
    this.session = session;
    this.config = config;
    this.rowMode = rowMode;
    this.forbidJsonb = forbidJsonb;
    this.mode = mode;
    this.table = table;
  }
  static [import_entity.entityKind] = "SQLiteAsyncRelationalQueryV2";
  /** @internal */
  mode;
  /** @internal */
  table;
  /** @internal */
  getSQL() {
    const query = this.dialect.buildRelationalQuery({
      schema: this.schema,
      tableNamesMap: this.tableNamesMap,
      table: this.table,
      tableConfig: this.tableConfig,
      queryConfig: this.config,
      tables: this.tables,
      mode: this.mode,
      jsonb: this.forbidJsonb ? import_sql.sql`json` : import_sql.sql`jsonb`
    });
    return query.sql;
  }
  /** @internal */
  _prepare(isOneTimeQuery = true) {
    const { query, builtQuery } = this._toSQL();
    return this.session[isOneTimeQuery ? "prepareOneTimeRelationalQuery" : "prepareRelationalQuery"](
      builtQuery,
      void 0,
      this.mode === "first" ? "get" : "all",
      (rawRows, mapColumnValue) => {
        const rows = rawRows.map((row) => (0, import_relations.mapRelationalRow)(row, query.selection, mapColumnValue, !this.rowMode));
        if (this.mode === "first") {
          return rows[0];
        }
        return rows;
      }
    );
  }
  prepare() {
    return this._prepare(false);
  }
  _getQuery() {
    const jsonb = this.forbidJsonb ? import_sql.sql`json` : import_sql.sql`jsonb`;
    const query = this.dialect.buildRelationalQuery({
      schema: this.schema,
      tableNamesMap: this.tableNamesMap,
      table: this.table,
      tableConfig: this.tableConfig,
      queryConfig: this.config,
      tables: this.tables,
      mode: this.mode,
      isNested: this.rowMode,
      jsonb
    });
    if (this.rowMode) {
      const jsonColumns = import_sql.sql.join(
        query.selection.map((s) => {
          return import_sql.sql`${import_sql.sql.raw(this.dialect.escapeString(s.key))}, ${s.selection ? import_sql.sql`${jsonb}(${import_sql.sql.identifier(s.key)})` : import_sql.sql.identifier(s.key)}`;
        }),
        import_sql.sql`, `
      );
      query.sql = import_sql.sql`select json_object(${jsonColumns}) as ${import_sql.sql.identifier("r")} from (${query.sql}) as ${import_sql.sql.identifier("t")}`;
    }
    return query;
  }
  _toSQL() {
    const query = this._getQuery();
    const builtQuery = this.dialect.sqlToQuery(query.sql);
    return { query, builtQuery };
  }
  toSQL() {
    return this._toSQL().builtQuery;
  }
  /** @internal */
  executeRaw() {
    if (this.mode === "first") {
      return this._prepare().get();
    }
    return this._prepare().all();
  }
  async execute() {
    return this.executeRaw();
  }
}
class SQLiteSyncRelationalQuery extends SQLiteRelationalQuery {
  static [import_entity.entityKind] = "SQLiteSyncRelationalQueryV2";
  sync() {
    return this.executeRaw();
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  RelationalQueryBuilder,
  SQLiteRelationalQuery,
  SQLiteSyncRelationalQuery
});
