// drizzle-orm/libsql/driver-core.js
import * as V1 from "../_relations.js";
import { entityKind } from "../entity.js";
import { DefaultLogger } from "../logger.js";
import { BaseSQLiteDatabase } from "../sqlite-core/db.js";
import { SQLiteAsyncDialect } from "../sqlite-core/dialect.js";
import { LibSQLSession } from "./session.js";
class LibSQLDatabase extends BaseSQLiteDatabase {
  static [entityKind] = "LibSQLDatabase";
  async batch(batch) {
    return this.session.batch(batch);
  }
}
function construct(client, config = {}) {
  const dialect = new SQLiteAsyncDialect({ casing: config.casing });
  let logger;
  if (config.logger === true) {
    logger = new DefaultLogger();
  } else if (config.logger !== false) {
    logger = config.logger;
  }
  let schema;
  if (config.schema) {
    const tablesConfig = V1.extractTablesRelationalConfig(
      config.schema,
      V1.createTableRelationsHelpers
    );
    schema = {
      fullSchema: config.schema,
      schema: tablesConfig.tables,
      tableNamesMap: tablesConfig.tableNamesMap
    };
  }
  const relations = config.relations;
  const session = new LibSQLSession(client, dialect, relations, schema, { logger }, void 0);
  const db = new LibSQLDatabase(
    "async",
    dialect,
    session,
    relations,
    schema
  );
  db.$client = client;
  return db;
}
export {
  LibSQLDatabase,
  construct
};
