// drizzle-orm/gel/driver.js
import { createClient } from "gel";
import * as V1 from "../_relations.js";
import { entityKind } from "../entity.js";
import { GelDatabase } from "../gel-core/db.js";
import { GelDialect } from "../gel-core/dialect.js";
import { DefaultLogger } from "../logger.js";
import { isConfig } from "../utils.js";
import { GelDbSession } from "./session.js";
class GelDriver {
  constructor(client, dialect, options = {}) {
    this.client = client;
    this.dialect = dialect;
    this.options = options;
  }
  static [entityKind] = "GelDriver";
  createSession(relations, schema) {
    return new GelDbSession(this.client, this.dialect, relations, schema, { logger: this.options.logger });
  }
}
class GelJsDatabase extends GelDatabase {
  static [entityKind] = "GelJsDatabase";
}
function construct(client, config = {}) {
  const dialect = new GelDialect({ casing: config.casing });
  let logger;
  if (config.logger === true) {
    logger = new DefaultLogger();
  } else if (config.logger !== false) {
    logger = config.logger;
  }
  let schema;
  if (config.schema) {
    const tablesConfig = V1.extractTablesRelationalConfig(config.schema, V1.createTableRelationsHelpers);
    schema = {
      fullSchema: config.schema,
      schema: tablesConfig.tables,
      tableNamesMap: tablesConfig.tableNamesMap
    };
  }
  const relations = config.relations;
  const driver = new GelDriver(client, dialect, { logger });
  const session = driver.createSession(relations, schema);
  const db = new GelJsDatabase(dialect, session, relations, schema);
  db.$client = client;
  return db;
}
function drizzle(...params) {
  if (typeof params[0] === "string") {
    const instance = createClient({ dsn: params[0] });
    return construct(instance, params[1]);
  }
  if (isConfig(params[0])) {
    const { connection, client, ...drizzleConfig } = params[0];
    if (client)
      return construct(client, drizzleConfig);
    const instance = createClient(connection);
    return construct(instance, drizzleConfig);
  }
  return construct(params[0], params[1]);
}
((drizzle2) => {
  function mock(config) {
    return construct({}, config);
  }
  drizzle2.mock = mock;
})(drizzle || (drizzle = {}));
export {
  GelDriver,
  GelJsDatabase,
  drizzle
};
