// drizzle-orm/xata-http/driver.js
import * as V1 from "../_relations.js";
import { entityKind } from "../entity.js";
import { DefaultLogger } from "../logger.js";
import { PgDatabase } from "../pg-core/db.js";
import { PgDialect } from "../pg-core/dialect.js";
import { XataHttpSession } from "./session.js";
class XataHttpDriver {
  constructor(client, dialect, options = {}) {
    this.client = client;
    this.dialect = dialect;
    this.options = options;
    this.initMappers();
  }
  static [entityKind] = "XataDriver";
  createSession(relations, schema) {
    return new XataHttpSession(this.client, this.dialect, relations, schema, {
      logger: this.options.logger
    });
  }
  initMappers() {
  }
}
class XataHttpDatabase extends PgDatabase {
  static [entityKind] = "XataHttpDatabase";
}
function drizzle(client, config = {}) {
  const dialect = new PgDialect({ casing: config.casing });
  let logger;
  if (config.logger === true) {
    logger = new DefaultLogger();
  } else if (config.logger !== false) {
    logger = config.logger;
  }
  let schema;
  if (config.schema) {
    const tablesConfig = V1.extractTablesRelationalConfig(config.schema, V1.createTableRelationsHelpers);
    schema = {
      fullSchema: config.schema,
      schema: tablesConfig.tables,
      tableNamesMap: tablesConfig.tableNamesMap
    };
  }
  const relations = config.relations;
  const driver = new XataHttpDriver(client, dialect, { logger });
  const session = driver.createSession(relations, schema);
  const db = new XataHttpDatabase(
    dialect,
    session,
    relations,
    schema
  );
  db.$client = client;
  return db;
}
export {
  XataHttpDatabase,
  XataHttpDriver,
  drizzle
};
